client/files_test.go

```diff
@@ -0,0 +1,280 @@
+      "type": "symlink",
+      "permissions": "777",
+      "last-modified": "2022-04-21T03:02:51Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+    },
+    {
+      "path": "/boot",
+      "name": "boot",
+      "type": "directory",
+      "permissions": "755",
+      "last-modified": "2022-08-16T10:23:25Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+    },
+    {
+      "path": "/proc",
+      "name": "proc",
+      "type": "directory",
+      "permissions": "555",
+      "last-modified": "2022-08-16T09:50:23Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+    },
+    {
+      "path": "/root",
+      "name": "root",
+      "type": "directory",
+      "permissions": "700",
+      "last-modified": "2022-08-17T11:55:09Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+    },
+    {
+      "path": "/swap.img",
+      "name": "swap.img",
+      "type": "file",
+      "size": 4100980736,
+      "permissions": "600",
+      "last-modified": "2022-05-14T18:10:12Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+    },
+    {
+      "path": "/tmp",
+      "name": "tmp",
+      "type": "directory",
+      "permissions": "777",
+      "last-modified": "2022-08-22T12:42:49Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+    },
+	{
+      "path": "/pebble.sock",
+      "name": "pebble.sock",
+      "type": "socket",
+      "permissions": "700",
+      "last-modified": "2022-08-22T12:42:49Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+	},
+	{
+      "path": "/dbus.sock",
+      "name": "dbus.sock",
+      "type": "named-pipe",
+      "permissions": "700",
+      "last-modified": "2022-08-22T12:42:49Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+	},
+	{
+      "path": "/tty0",
+      "name": "tty0",
+      "type": "device",
+      "permissions": "700",
+      "last-modified": "2022-08-22T12:42:49Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+	},
+	{
+      "path": "/irreg",
+      "name": "irreg",
+      "type": "sfdeljknesv",
+      "permissions": "700",
+      "last-modified": "2022-08-22T12:42:49Z",
+      "user-id": 0,
+      "user": "root",
+      "group-id": 0,
+      "group": "root"
+	}
+  ]
+}`
+	result, err := cs.cli.ListFiles(&client.ListFilesOptions{
+		Path: "/",
+	})
+	c.Assert(err, IsNil)
+	c.Assert(result, HasLen, 10)
+}
+
+func (cs *clientSuite) TestListDirectoryItself(c *C) {
+	cs.rsp = `{
+  "type": "sync",
+  "status-code": 200,
+  "status": "OK",
+  "result": [
+    {
+      "path": "/bin",
+      "name": "bin",
+      "type": "symlink",
+      "permissions": "777",
+      "last-modified": "2022-04-21T03:02:51Z",
+      "user-id": 0,
```

Comment:
We should probably test a value other than 0 for user-id/group-id, because 0 is the zero value so could happen "by accident". (Or we can test it above in `TestListFiles`.)

Reply:
Actually, I was thinking that the API can return a `nil` uid/gid, but `client.FileInfo` actually uses `int` and not `*int` for those. Should we keep using `*int` to allow for `nil` values on those fields?

https://github.com/canonical/pebble/blob/4855fa64b0a3e89fca33958306c20dbe19ab39f1/client/files.go#L191-L196

I don't think `userID` and `groupID` should be 0 (a perfectly valid uid/gid) if the API returns `nil` for those. For the `size` attribute things are a bit different because we need to implement `os.FileInfo.Size() int64` and I don't know if using a sentinel value like -1 works in this context.

What do you think?

Reply:
Good point. We should probably do `*int` for UserID and GroupID to match the API and semantics.

I think size is fine as is (no need to return -1) because it's explicitly marked as implementation-defined if type!=file.